File: javascript/src/fingerprint.js
/**
 * Hardware fingerprinting for device-bound encryption.
 *
 * Collects platform-specific hardware identifiers to create device-bound keys.
 * Adapted from SaaSClient-SideLicensingSystem.
 *
 * @module fingerprint
 */

const { execSync } = require('child_process');
const crypto = require('crypto');
const os = require('os');

/**
 * Fingerprint collection mode.
 * @enum {string}
 */
const FingerprintMode = {
    NONE: 'none',
    MOTHERBOARD: 'motherboard',
    CPU: 'cpu',
    COMBINED: 'combined'
};

/**
 * Collect hardware fingerprint.
 *
 * @param {string} mode - Fingerprint mode
 * @returns {string} Fingerprint string (MD5 hex), or empty for NONE
 * @throws {Error} If fingerprint unavailable
 *
 * @example
 * const fp = collectFingerprint(FingerprintMode.COMBINED);
 * console.log('Device fingerprint:', fp);
 */
function collectFingerprint(mode) {
    if (mode === FingerprintMode.NONE) {
        return '';
    }

    if (mode === FingerprintMode.MOTHERBOARD) {
        return getMotherboardSerial();
    }

    if (mode === FingerprintMode.CPU) {
        return getCpuId();
    }

    if (mode === FingerprintMode.COMBINED) {
        const components = [];

        try {
            components.push(getMotherboardSerial());
        } catch (e) {
            // Continue without motherboard
        }

        try {
            components.push(getCpuId());
        } catch (e) {
            // Continue without CPU
        }

        if (components.length === 0) {
            throw new Error('Hardware fingerprint unavailable');
        }

        // Create MD5 hash of combined components
        const combined = components.join('-');
        return crypto.createHash('md5').update(combined).digest('hex');
    }

    throw new Error(`Unknown fingerprint mode: ${mode}`);
}

/**
 * Get motherboard serial number (platform-specific).
 *
 * @returns {string} Motherboard serial
 * @throws {Error} If unavailable
 */
function getMotherboardSerial() {
    const platform = os.platform();

    if (platform === 'win32') {
        try {
            const output = execSync('wmic baseboard get serialnumber /value', {
                encoding: 'utf8',
                timeout: 5000
            });

            for (const line of output.split('\n')) {
                if (line.startsWith('SerialNumber=')) {
                    const serial = line.replace('SerialNumber=', '').trim();
                    if (serial && serial !== 'To be filled by O.E.M.') {
                        return serial;
                    }
                }
            }
        } catch (e) {
            // Fall through
        }
    } else if (platform === 'linux') {
        // Try DMI sysfs first
        try {
            const fs = require('fs');
            const serial = fs.readFileSync('/sys/class/dmi/id/board_serial', 'utf8').trim();
            if (serial && serial !== 'To be filled by O.E.M.') {
                return serial;
            }
        } catch (e) {
            // Fall through to dmidecode
        }

        // Fallback to dmidecode
        try {
            const output = execSync('dmidecode -s baseboard-serial-number', {
                encoding: 'utf8',
                timeout: 5000
            });
            const serial = output.trim();
            if (serial && serial !== 'To be filled by O.E.M.') {
                return serial;
            }
        } catch (e) {
            // Fall through
        }
    } else if (platform === 'darwin') {
        try {
            const output = execSync('system_profiler SPHardwareDataType', {
                encoding: 'utf8',
                timeout: 5000
            });

            for (const line of output.split('\n')) {
                if (line.includes('Serial Number')) {
                    const parts = line.split(':');
                    if (parts.length >= 2) {
                        return parts[1].trim();
                    }
                }
            }
        } catch (e) {
            // Fall through
        }
    }

    throw new Error('Motherboard serial number unavailable');
}

/**
 * Get CPU identifier (platform-specific).
 *
 * @returns {string} CPU ID
 * @throws {Error} If unavailable
 */
function getCpuId() {
    const platform = os.platform();

    if (platform === 'win32') {
        try {
            const output = execSync('wmic cpu get ProcessorId /value', {
                encoding: 'utf8',
                timeout: 5000
            });

            for (const line of output.split('\n')) {
                if (line.startsWith('ProcessorId=')) {
                    const cpuId = line.replace('ProcessorId=', '').trim();
                    if (cpuId) {
                        return cpuId;
                    }
                }
            }
        } catch (e) {
            // Fall through
        }
    } else if (platform === 'linux') {
        try {
            const fs = require('fs');
            const content = fs.readFileSync('/proc/cpuinfo', 'utf8');

            for (const line of content.split('\n')) {
                if (line.startsWith('processor') && line.includes('0')) {
                    // Use first processor as identifier (hashed)
                    return crypto.createHash('md5').update(line).digest('hex');
                }
            }
        } catch (e) {
            // Fall through
        }
    } else if (platform === 'darwin') {
        try {
            const output = execSync('sysctl -n machdep.cpu.brand_string', {
                encoding: 'utf8',
                timeout: 5000
            });
            const cpuInfo = output.trim();
            if (cpuInfo) {
                return crypto.createHash('md5').update(cpuInfo).digest('hex');
            }
        } catch (e) {
            // Fall through
        }
    }

    throw new Error('CPU identifier unavailable');
}

module.exports = {
    FingerprintMode,
    collectFingerprint
};

File: java/src/main/java/ai/guard8/shield/Fingerprint.java
package ai.guard8.shield;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.List;

/**
 * Hardware fingerprinting for device-bound encryption.
 *
 * <p>Collects platform-specific hardware identifiers to create device-bound keys.
 * Adapted from SaaSClient-SideLicensingSystem.
 *
 * <p>Example:
 * <pre>{@code
 * String fp = Fingerprint.collect(FingerprintMode.COMBINED);
 * Shield shield = Shield.withFingerprint("password", "github.com", FingerprintMode.COMBINED);
 * }</pre>
 */
public class Fingerprint {

    /**
     * Fingerprint collection mode.
     */
    public enum FingerprintMode {
        /** No fingerprinting (backward compatible) */
        NONE,
        /** Motherboard serial only */
        MOTHERBOARD,
        /** CPU identifier only */
        CPU,
        /** Motherboard + CPU (recommended) */
        COMBINED
    }

    /**
     * Collect hardware fingerprint.
     *
     * @param mode Fingerprint mode
     * @return Fingerprint string (MD5 hex), or empty for NONE
     * @throws Exception If hardware identifiers cannot be collected
     */
    public static String collect(FingerprintMode mode) throws Exception {
        if (mode == FingerprintMode.NONE) {
            return "";
        }

        if (mode == FingerprintMode.MOTHERBOARD) {
            return getMotherboardSerial();
        }

        if (mode == FingerprintMode.CPU) {
            return getCpuId();
        }

        if (mode == FingerprintMode.COMBINED) {
            List<String> components = new ArrayList<>();

            try {
                components.add(getMotherboardSerial());
            } catch (Exception e) {
                // Continue without motherboard
            }

            try {
                components.add(getCpuId());
            } catch (Exception e) {
                // Continue without CPU
            }

            if (components.isEmpty()) {
                throw new Exception("Hardware fingerprint unavailable");
            }

            // Create MD5 hash of combined components
            String combined = String.join("-", components);
            return md5(combined);
        }

        throw new IllegalArgumentException("Unknown fingerprint mode: " + mode);
    }

    /**
     * Get motherboard serial number (platform-specific).
     */
    private static String getMotherboardSerial() throws Exception {
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("win")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"wmic", "baseboard", "get", "serialnumber", "/value"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.startsWith("SerialNumber=")) {
                        String serial = line.replace("SerialNumber=", "").trim();
                        if (!serial.isEmpty() && !serial.equals("To be filled by O.E.M.")) {
                            return serial;
                        }
                    }
                }
            }
        } else if (os.contains("linux")) {
            // Try DMI sysfs first
            try {
                String serial = new String(Files.readAllBytes(
                    Paths.get("/sys/class/dmi/id/board_serial")
                )).trim();
                if (!serial.isEmpty() && !serial.equals("To be filled by O.E.M.")) {
                    return serial;
                }
            } catch (Exception e) {
                // Fall through to dmidecode
            }

            // Fallback to dmidecode
            Process process = Runtime.getRuntime().exec(
                new String[]{"dmidecode", "-s", "baseboard-serial-number"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String serial = reader.readLine();
                if (serial != null) {
                    serial = serial.trim();
                    if (!serial.isEmpty() && !serial.equals("To be filled by O.E.M.")) {
                        return serial;
                    }
                }
            }
        } else if (os.contains("mac")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"system_profiler", "SPHardwareDataType"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.contains("Serial Number")) {
                        String[] parts = line.split(":");
                        if (parts.length >= 2) {
                            return parts[1].trim();
                        }
                    }
                }
            }
        }

        throw new Exception("Motherboard serial number unavailable");
    }

    /**
     * Get CPU identifier (platform-specific).
     */
    private static String getCpuId() throws Exception {
        String os = System.getProperty("os.name").toLowerCase();

        if (os.contains("win")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"wmic", "cpu", "get", "ProcessorId", "/value"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.startsWith("ProcessorId=")) {
                        String cpuId = line.replace("ProcessorId=", "").trim();
                        if (!cpuId.isEmpty()) {
                            return cpuId;
                        }
                    }
                }
            }
        } else if (os.contains("linux")) {
            String content = new String(Files.readAllBytes(Paths.get("/proc/cpuinfo")));
            for (String line : content.split("\n")) {
                if (line.startsWith("processor") && line.contains("0")) {
                    // Use first processor as identifier (hashed)
                    return md5(line);
                }
            }
        } else if (os.contains("mac")) {
            Process process = Runtime.getRuntime().exec(
                new String[]{"sysctl", "-n", "machdep.cpu.brand_string"}
            );

            try (BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()))) {
                String cpuInfo = reader.readLine();
                if (cpuInfo != null) {
                    cpuInfo = cpuInfo.trim();
                    if (!cpuInfo.isEmpty()) {
                        return md5(cpuInfo);
                    }
                }
            }
        }

        throw new Exception("CPU identifier unavailable");
    }

    /**
     * MD5 hash helper.
     */
    private static String md5(String input) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] digest = md.digest(input.getBytes());

        StringBuilder hex = new StringBuilder();
        for (byte b : digest) {
            hex.append(String.format("%02x", b));
        }
        return hex.toString();
    }
}
